jax_md/elasticity.py

```diff
@@ -0,0 +1,582 @@
+  
+  rank = len(M.shape)
+  if not (rank==1 or rank==2):
+    raise AssertionError('T must have rank 1 or 2')
+
+  def mandel_index(i,j):
+    return lax.cond(i==j, lambda ij: ij[0], lambda ij: DIM-ij[0]-ij[1], (i,j))
+
+  if DIM == 3:
+    dimension = 2
+  else:
+    dimension = 3
+
+  tensor_range = jnp.arange(dimension)
+  _, weight = _get_mandel_mapping_weight(dimension)
+
+  if rank == 1:
+    def extract(i,j):
+      idx = mandel_index(i,j)
+      return M[idx] / weight[idx]
+    T = vmap(vmap(extract, in_axes=(0,None)), in_axes=(None,0))(tensor_range, tensor_range)
+  else:
+    def extract(i,j,k,l):
+      idx0 = mandel_index(i,j)
+      idx1 = mandel_index(k,l)
+      return M[idx0,idx1] / (weight[idx0] * weight[idx1])
+    T = vmap(vmap(vmap(vmap(extract, in_axes=(0,None,None,None)), in_axes=(None,0,None,None)), in_axes=(None,None,0,None)), in_axes=(None,None,None,0))(tensor_range, tensor_range, tensor_range, tensor_range)
+  return T
+
+
+
+@partial(jit,static_argnums=(1,))
+def _extract_elements(C, as_dict):
+  """ Convert an elastic modulus tensor into a list of 6 (21) unique elements
+        in 2 (3) dimensions
+      
+      in 2d, these are:
+      cxxxx,cyyyy,cxyxy,cxxyy,cxxxy,cyyxy
+
+      in 3d, these are:
+      cxxxx,cyyyy,czzzz,cyzyz,cxzxz,cxyxy,cyyzz,cxxzz,cxxyy,cxxyz,cxxxz,cxxxy,cyyyz,cyyxz,cyyxy,czzyz,czzxz,czzxy,cyzxz,cyzxy,cxzxy
+
+  Args:
+    C: A previously calculated elastic modulus tensor represented as an 
+      array of shape (spatial_dimension,spatial_dimension,spatial_dimension,
+      spatial_dimension), where spatial_dimension is either 2 or 3. C must 
+      satisfy both the major and minor symmetries, but this is not checked.
+    as_dict: boolean. If true, return a dictionary with interpretable keys.
+      If false, return an array that follows an internal convention.
+  """
+  if C.shape[0] == 2:
+    indices = ( [0, 1, 0, 0, 0, 0],
+                [0, 1, 1, 0, 0, 1],
+                [0, 1, 0, 1, 0, 1],
+                [0, 1, 1, 1, 1, 1])
+    clist = C[ indices ] 
+    if as_dict==True:
+      names = ['cxxxx','cyyyy','cxyxy','cxxyy','cxxxy','cyyxy']
+      return dict(zip(names, clist))
+    else:
+      return clist
+
+  elif C.shape[0] == 3:
+    indices = ( [0, 1, 2, 1, 0, 0, 1, 0, 0, 0, 0, 0, 1, 0, 0, 1, 0, 0, 0, 0, 0],
+                [0, 1, 2, 2, 2, 1, 1, 0, 0, 0, 0, 0, 1, 2, 1, 2, 2, 1, 2, 1, 1],
+                [0, 1, 2, 1, 0, 0, 2, 2, 1, 1, 0, 0, 1, 1, 1, 2, 2, 2, 1, 1, 0],
+                [0, 1, 2, 2, 2, 1, 2, 2, 1, 2, 2, 1, 2, 1, 1, 2, 2, 2, 2, 2, 2])
+    clist = C[ indices ] 
+    if as_dict:
+      names = ['cxxxx','cyyyy','czzzz','cyzyz','cxzxz','cxyxy','cyyzz','cxxzz','cxxyy','cxxyz','cxxxz','cxxxy','cyyyz','cyyxz','cyyxy','czzyz','czzxz','czzxy','cyzxz','cyzxy','cxzxy']
+      return dict(zip(names, clist))
+    else:
+      return clist
+  else:
+    raise AssertionError('C has wrong shape')
+
+def extract_elements(C: Array) -> Dict:
```

Comment:
redefinition of unused 'extract elements' from line 269. 
It seems to me that the first 'extract elements' is the same function as '_extract elements' and can therefor be removed.